feat: AccountBalanceQuery deprecation and mirror node migration proposals (#275)
hiero-ledger/sdk-collaboration-hub: proposals/account-balance-query-mirror-node-migration.md

# AccountBalanceQuery Mirror Node Migration

## Summary

`AccountBalanceQuery` currently retrieves HBAR and token balances by issuing a free gRPC call to a consensus node (`CryptoService/cryptoGetBalance`). The Hedera network is deprecating this endpoint. This proposal introduces a replacement query, `MirrorNodeAccountBalanceQuery`, that fetches the same data from the mirror node REST API (`GET /api/v1/accounts/{idOrAliasOrEvmAddress}`), and simultaneously deprecates `AccountBalanceQuery`.

> The new class follows the naming and structural convention already established in the SDKs for mirror node REST queries (`MirrorNodeContractCallQuery`, `MirrorNodeContractEstimateQuery`). The existing `AccountBalanceQuery` is marked deprecated but not removed, giving developers a migration window. A future `BlockNodeAccountBalanceQuery` is noted as a named placeholder for when block node infrastructure is available across the network.

Reviewer: **Conflict with the companion proposal is unstated.** This doc keeps `AccountBalanceQuery` working (Test 12), while `account-balance-query-deprecation.md` makes `execute()` throw immediately (its Tests 2–3). Both cannot be adopted as written, and this doc never links back to the other. State the relationship explicitly:

````suggestion
The new class follows the naming and structural convention already established in the SDKs for mirror node REST queries (`MirrorNodeContractCallQuery`, `MirrorNodeContractEstimateQuery`). The existing `AccountBalanceQuery` is marked deprecated but not removed, giving developers a migration window. A future `BlockNodeAccountBalanceQuery` is noted as a named placeholder for when block node infrastructure is available across the network.

**Relationship to the companion proposal:** The companion [AccountBalanceQuery Deprecation](./account-balance-query-deprecation.md) proposal makes `execute()` fail immediately with a hard error, while this proposal keeps `AccountBalanceQuery` fully functional during a migration window. The two are mutually exclusive strategies for retiring the consensus-node balance path; only one of them can be adopted.
````

- [Mirror node REST docs: GET /api/v1/contracts/{id}](https://docs.hedera.com/hedera/sdks-and-apis/rest-api#contracts)
- [Docs: Get Account Token Balance](https://docs.hedera.com/native/tokens/get-balance#get-account-token-balance)

---

## New APIs

### `MirrorNodeAccountBalanceQuery`

A new standalone query class that fetches account or contract balance from the mirror node REST API. Does not extend the base `Query` class (which carries consensus-node gRPC machinery). Follows the same pattern as `MirrorNodeContractCallQuery`.

```
MirrorNodeAccountBalanceQuery {
    AccountId | null    accountId
    ContractId | null   contractId

    MirrorNodeAccountBalanceQuery setAccountId(accountId: AccountId | string)
    MirrorNodeAccountBalanceQuery setContractId(contractId: ContractId | string)

    Promise<AccountBalance> execute(client: Client)
}
```

`setAccountId` accepts any form `AccountId.fromString(...)` accepts: `shard.realm.num`, EVM address (`0x...`), or public key alias. The mirror node natively resolves all three, which is an improvement over the consensus node path.

> `setContractId` routes to `GET /api/v1/contracts/{contractId}` instead; see Internal Changes.

Reviewer: **Meta-language rewrite.** `Type | null` fields, `AccountId | string` parameter unions, and `Promise<>` return types are not meta-language syntax (`guides/api-guideline.md`): nullability is `@@nullable`, async is `@@async`, and the error cases defined under Response Codes belong in `@@throws(...)`. The accountId/contractId exclusivity implied by the routing rules is exactly `@@oneOrNoneOf`.

````suggestion
```
@@oneOrNoneOf(accountId, contractId)
MirrorNodeAccountBalanceQuery {
    @@nullable accountId: AccountId
    @@nullable contractId: ContractId

    MirrorNodeAccountBalanceQuery setAccountId(accountId: AccountId)
    MirrorNodeAccountBalanceQuery setContractId(contractId: ContractId)

    @@async
    @@throws(invalid-argument-error, not-found-error)
    AccountBalance execute(client: Client)
}
```

`accountId` accepts any form the SDK's `AccountId` string parsing accepts: `shard.realm.num`, EVM address
(`0x...`), or public key alias. The mirror node natively resolves all three, which is an improvement over the
consensus node path.

Setting `contractId` routes the query to `GET /api/v1/contracts/{contractId}` instead; see Internal Changes.
````

### Future placeholder: `BlockNodeAccountBalanceQuery`

A `BlockNodeAccountBalanceQuery` class is reserved for a future proposal once block node infrastructure is broadly available. Its public API is expected to mirror `MirrorNodeAccountBalanceQuery`. No implementation is defined here.

---

> ## Updated APIs
> 
> ### `AccountBalanceQuery` — deprecated
> 
> `AccountBalanceQuery` is marked `@deprecated` across all SDKs. Its implementation and behavior are **unchanged** — it continues to issue gRPC calls to consensus nodes for the duration of the deprecation window. No fields or methods are removed. SDKs should emit a deprecation warning on construction or first `execute()` call pointing developers to `MirrorNodeAccountBalanceQuery`.
> 
> ```
> @deprecated Use MirrorNodeAccountBalanceQuery instead.
> AccountBalanceQuery {
>     AccountId | null    accountId     // unchanged
>     ContractId | null   contractId    // unchanged
> 
>     AccountBalanceQuery setAccountId(accountId: AccountId | string)
>     AccountBalanceQuery setContractId(contractId: ContractId | string)
> 
>     Promise<AccountBalance> execute(client: Client)
> }
> ```
> 
> ### `AccountBalance` — `tokenDecimals` deprecated
> 
> `tokenDecimals` is retained on `AccountBalance` so that existing code using `AccountBalanceQuery` continues to compile and run unchanged during the deprecation window. When using the new `MirrorNodeAccountBalanceQuery`, `tokenDecimals` will always be empty (the mirror node account balance endpoint does not return decimals). SDKs should mark it `@deprecated`.
> 
> ```
> AccountBalance {
>     Hbar            hbars           // unchanged
>     TokenBalanceMap tokens          // unchanged
>     @deprecated
>     TokenDecimalMap tokenDecimals   // empty when returned by MirrorNodeAccountBalanceQuery
> }
> ```

Reviewer: **Meta-language rewrite** — same issues as the New APIs block (`Type | null`, `| string` unions, `Promise<>`), and `@deprecated` (single `@`, JSDoc style) is not a meta-language annotation; the guideline defines none for deprecation, so it is expressed in prose.

````suggestion
```
@@oneOrNoneOf(accountId, contractId)
AccountBalanceQuery {
    @@nullable accountId: AccountId     // unchanged
    @@nullable contractId: ContractId   // unchanged

    AccountBalanceQuery setAccountId(accountId: AccountId)
    AccountBalanceQuery setContractId(contractId: ContractId)

    @@async
    AccountBalance execute(client: Client)
}
```

The meta-language defines no deprecation annotation; each SDK marks the class deprecated in its idiomatic way
(e.g., JSDoc `@deprecated`, Java `@Deprecated`, Rust `#[deprecated]`).

### `AccountBalance` — `tokenDecimals` deprecated

`tokenDecimals` is retained on `AccountBalance` so that existing code using `AccountBalanceQuery` continues to
compile and run unchanged during the deprecation window. When using the new `MirrorNodeAccountBalanceQuery`,
`tokenDecimals` will always be empty (the mirror node account balance endpoint does not return decimals). SDKs
should mark it deprecated in their idiomatic way.

```
AccountBalance {
    hbars: Hbar                     // unchanged
    tokens: TokenBalanceMap         // unchanged
    tokenDecimals: TokenDecimalMap  // deprecated — empty when returned by MirrorNodeAccountBalanceQuery
}
```
````

   - Parse `balance.tokens[]` → `AccountBalance.tokens` (TokenBalanceMap), following `next` pagination links until exhausted.
   - Leave `AccountBalance.tokenDecimals` empty.

2. If `contractId` is set:
   - Issue `GET /api/v1/contracts/{contractId}`.
   - Parse `balance` (tinybars) → `AccountBalance.hbars`.
   - Return an empty `TokenBalanceMap` (the mirror node contracts endpoint does not expose token balances).

**Mirror response shape (account path):**

```json
{
  "balance": {
    "balance": 123456789,
    "tokens": [
      { "token_id": "0.0.12345", "balance": 500 },
      { "token_id": "0.0.67890", "balance": 100 }
    ]
  },
  "links": { "next": "/api/v1/accounts/0.0.12345/tokens?limit=25&..." }
}
```

### Why `tokenDecimals` cannot be populated

The mirror node `GET /api/v1/accounts/{id}` response does not include token decimals. Retrieving them would require one `GET /api/v1/tokens/{tokenId}` call per token held — an unbounded number of extra calls for accounts with large token portfolios. `tokenDecimals` is therefore empty when using `MirrorNodeAccountBalanceQuery`. Developers who need decimals should use a `TokenInfoQuery` (or equivalent) per token.

### Pagination

> The `balance.tokens` array in the mirror node account response is paginated. `MirrorNodeAccountBalanceQuery` must follow `links.next` until it is absent, accumulating all token entries before constructing the `TokenBalanceMap`. This matches the pagination behavior of `AddressBookQueryWeb`.

Reviewer: **Wrong pagination endpoint + JS-specific internals.** (1) `links.next` on `GET /api/v1/accounts/{id}` paginates the account's *transactions*, not `balance.tokens` — complete token balances come from `GET /api/v1/accounts/{id}/tokens` (which the example `next` link already points to; the embedded `balance.tokens` array is only a truncated preview). (2) This spec is normative for all 8 SDKs, so `fetch` and `client.mirrorRestApiBaseUrl` should be described generically with JS cited as precedent.

````suggestion
### `MirrorNodeAccountBalanceQuery` implementation

The class does not extend `Query`. It uses the SDK's HTTP abstraction and the mirror node REST base URL from the
client's mirror network configuration, following the structure of the existing mirror node REST queries
(precedent in the JS SDK: `MirrorNodeContractCallQuery`, `FeeEstimateQuery`).

**Request routing:**

1. If `accountId` is set:
   - Issue `GET /api/v1/accounts/{accountId}` where `accountId` is the string form of the `AccountId`
     (shard.realm.num, EVM address, or alias — all accepted by the mirror node).
   - Parse `balance.balance` (tinybars) → `AccountBalance.hbars`.
   - Issue `GET /api/v1/accounts/{accountId}/tokens` and follow `links.next` until it is absent, accumulating
     all entries into `AccountBalance.tokens` (`TokenBalanceMap`). The `balance.tokens` array embedded in the
     account response is a truncated preview and must not be used as the source of token balances.
   - Leave `AccountBalance.tokenDecimals` empty.

2. If `contractId` is set:
   - Issue `GET /api/v1/contracts/{contractId}`.
   - Parse `balance` (tinybars) → `AccountBalance.hbars`.
   - Return an empty `TokenBalanceMap` (the mirror node contracts endpoint does not expose token balances).

**Mirror response shape (tokens path):**

```json
{
  "tokens": [
    { "token_id": "0.0.12345", "balance": 500 },
    { "token_id": "0.0.67890", "balance": 100 }
  ],
  "links": { "next": "/api/v1/accounts/0.0.12345/tokens?limit=25&..." }
}
```

### Why `tokenDecimals` cannot be populated

The mirror node `GET /api/v1/accounts/{id}` response does not include token decimals. Retrieving them would require one `GET /api/v1/tokens/{tokenId}` call per token held — an unbounded number of extra calls for accounts with large token portfolios. `tokenDecimals` is therefore empty when using `MirrorNodeAccountBalanceQuery`. Developers who need decimals should use a `TokenInfoQuery` (or equivalent) per token.

### Pagination

Token balances are paginated by the mirror node. `MirrorNodeAccountBalanceQuery` must follow `links.next` on
`GET /api/v1/accounts/{id}/tokens` until it is absent, accumulating all token entries before constructing the
`TokenBalanceMap`. Note that `links.next` on the account endpoint itself paginates the account's transactions,
not its token balances.
````

12. Given a call to the deprecated `AccountBalanceQuery`, when it is executed, then it still returns a correct result via the consensus node gRPC path (no behavioral regression during the deprecation window).

### TCK

The above tests should each have a corresponding issue created in `hiero-ledger/hiero-sdk-tck` and linked back to this design doc. Tests 3, 4, and 10 exercise capabilities not covered by the legacy consensus-node path and should be prioritized.

---

## SDK Example

### HBAR and token balances (new API)

```javascript
import { MirrorNodeAccountBalanceQuery, Client } from "@hiero-ledger/sdk";

const client = Client.forTestnet();

const balance = await new MirrorNodeAccountBalanceQuery()
    .setAccountId("0.0.12345")
    .execute(client);

console.log(`HBAR balance: ${balance.hbars.toString()}`);

for (const [tokenId, amount] of balance.tokens) {
    console.log(`Token ${tokenId.toString()}: ${amount.toString()}`);
}
```

### Lookup by EVM address (new capability)

```javascript
const balance = await new MirrorNodeAccountBalanceQuery()

>     .setAccountId("0x00000000000000000000000000000000000bc614e")

Reviewer: Invalid EVM address — 41 hex characters after `0x`; an EVM address is exactly 40.

````suggestion
    .setAccountId("0x0000000000000000000000000000000000bc614e")
````

```

`accountId` accepts any form the SDK's `AccountId` string parsing accepts: `shard.realm.num`, EVM address
(`0x...`), or public key alias. The mirror node natively resolves all three, which is an improvement over the
consensus node path.

Setting `contractId` routes the query to `GET /api/v1/contracts/{contractId}` instead; see Internal Changes.

### Future placeholder: `BlockNodeAccountBalanceQuery`

A `BlockNodeAccountBalanceQuery` class is reserved for a future proposal once block node infrastructure is broadly available. Its public API is expected to mirror `MirrorNodeAccountBalanceQuery`. No implementation is defined here.

---

## Updated APIs

### `AccountBalanceQuery` — deprecated

`AccountBalanceQuery` is marked deprecated across all SDKs. Its implementation and behavior are **unchanged** — it continues to issue gRPC calls to consensus nodes for the duration of the deprecation window. No fields or methods are removed. SDKs should emit a deprecation warning on construction or first `execute()` call pointing developers to `MirrorNodeAccountBalanceQuery`.

```
@@oneOrNoneOf(accountId, contractId)
AccountBalanceQuery {
    @@nullable accountId: AccountId     // unchanged

>     @@nullable contractId: ContractId   // unchanged

Reviewer 2: Is it necessary to have both fields? The response is `AccountBalance`, not `AccountOrPossiblyContractBalance`. Internally, regardless of contract or account it will get routed to the same MN endpoint. A contract is also an account so we could just keep it as accountId only and add docs for acknowledging it also works for contracts.

    AccountBalanceQuery setAccountId(accountId: AccountId)
    AccountBalanceQuery setContractId(contractId: ContractId)

    @@async
    AccountBalance execute(client: Client)
}
```

The meta-language defines no deprecation annotation; each SDK marks the class deprecated in its idiomatic way
(e.g., JSDoc `@deprecated`, Java `@Deprecated`, Rust `#[deprecated]`).

### `AccountBalance` — `tokenDecimals` deprecated

`tokenDecimals` is retained on `AccountBalance` so that existing code using `AccountBalanceQuery` continues to
compile and run unchanged during the deprecation window. When using the new `MirrorNodeAccountBalanceQuery`,
`tokenDecimals` will always be empty (the mirror node account balance endpoint does not return decimals). SDKs
should mark it deprecated in their idiomatic way.

```
AccountBalance {
    hbars: Hbar                     // unchanged
    tokens: TokenBalanceMap         // unchanged

>     tokenDecimals: TokenDecimalMap  // deprecated — empty when returned by MirrorNodeAccountBalanceQuery

Reviewer 2: We should not return tokens or tokenDecimals. There is no token balances on `/api/v1/accounts/{id}` and `/api/v1/balances?account.id=` does not paginate. We should not attempt to reuse an existing `AccountBalance` object and instead create a custom `MirrorNodeAccountBalance`.

}
```

---

## Internal Changes

### `MirrorNodeAccountBalanceQuery` implementation

The class does not extend `Query`. It uses the SDK's HTTP abstraction and the mirror node REST base URL from the
client's mirror network configuration, following the structure of the existing mirror node REST queries
(precedent in the JS SDK: `MirrorNodeContractCallQuery`, `FeeEstimateQuery`).

**Request routing:**

1. If `accountId` is set:
   - Issue `GET /api/v1/accounts/{accountId}` where `accountId` is the string form of the `AccountId`

>      (shard.realm.num, EVM address, or alias — all accepted by the mirror node).

Reviewer 2: `/api/v1/balances?account.id=` is more performant than this endpoint so I'd prefer to use that. It also supports alias or evm address as well as contracts.

   - Parse `balance.balance` (tinybars) → `AccountBalance.hbars`.
   - Issue `GET /api/v1/accounts/{accountId}/tokens` and follow `links.next` until it is absent, accumulating
     all entries into `AccountBalance.tokens` (`TokenBalanceMap`). The `balance.tokens` array embedded in the

>      account response is a truncated preview and must not be used as the source of token balances.

Reviewer 2: We definitely cannot do this. There can be an unbounded amount of tokens per account. Even if there's not, this will be a DDoS against mirror node for querying every token balance regardless if they need it.

   - Leave `AccountBalance.tokenDecimals` empty.

2. If `contractId` is set:
   - Issue `GET /api/v1/contracts/{contractId}`.
   - Parse `balance` (tinybars) → `AccountBalance.hbars`.

>    - Return an empty `TokenBalanceMap` (the mirror node contracts endpoint does not expose token balances).

Reviewer 2: We should not need to call /api/v1/contracts as both /api/v1/accounts/{id} and balances endpoint support a contract ID

### Pagination

Token balances are paginated by the mirror node. `MirrorNodeAccountBalanceQuery` must follow `links.next` on
`GET /api/v1/accounts/{id}/tokens` until it is absent, accumulating all token entries before constructing the
`TokenBalanceMap`. Note that `links.next` on the account endpoint itself paginates the account's transactions,
not its token balances.

### Eventual consistency

The mirror node reflects network state with a small lag (typically seconds). `AccountBalanceQuery` (consensus node) returns data current as of the latest handled block. Applications that require immediate post-transaction balance confirmation should allow for this lag when using `MirrorNodeAccountBalanceQuery`. This should be documented in SDK release notes.

### Free query — no change

Both `AccountBalanceQuery` and `MirrorNodeAccountBalanceQuery` are free. No payment logic is involved in either path.

### Response Codes

No consensus node response codes apply to `MirrorNodeAccountBalanceQuery`. Mirror node HTTP errors:

- `400 Bad Request` — invalid ID format. Surface as an SDK-appropriate error with a clear message. Do not retry.

> - `404 Not Found` — account or contract does not exist. Surface as an SDK-appropriate error. Do not retry.

Reviewer 2: If we use balances endpoint, it will not return 404 but empty list
